cylc/flow/task_pool.py

```diff
@@ -85,6 +86,154 @@
 Pool = Dict['PointBase', Dict[str, TaskProxy]]
 
 
+class TaskHoldMgr:
+    """Hold/release logic for active and future tasks.
+
+    Active tasks (i.e., task proxies in the pool):
+    - hold/release with --flow=n, or (by default) regardless of flow.
+
+    Future tasks (point/name):
+    - flagg for future hold, or unflag, with --flow=n, or (by default)
+      regardless of flow.
+
+    Note this class doesn't yet handle workflow hold point.
+
+    """
+    def __init__(
+        self,
+        workflow_db_mgr: 'WorkflowDatabaseManager',
+        data_store_mgr: 'DataStoreMgr',
+    ):
+        # (name, point): flow
+        self.hold: Dict[Tuple[str, 'PointBase'], Optional[int]] = {}
+        # flow may be None: hold future task regardless of its flow number.
+        # NOTE: RHS could be a set  of flow numbers, meaning future-hold same
+        # task in multiple specific flows. But those instances can't coexist in
+        # the pool so serially holding them is probably fine.
+        self.data_store_mgr = data_store_mgr
+        self.db_mgr = workflow_db_mgr
+
+    def _flatten(self):
+        # possibly-temporary conversion to old-style flat set
+        result = set()
+        for (name, point), flow in self.hold.items():
+            result.add((name, point, flow))
+        return result
+
+    def _update_stores(
+        self,
+        itask: Union[TaskProxy, Tuple[str, 'PointBase', bool]]
+    ):
+        """Update datastore and database."""
+        self.db_mgr.put_tasks_to_hold(self._flatten())
+        self.data_store_mgr.delta_task_held(itask)
+        LOG.debug(f"Tasks to hold {self.hold}")
+
+    def load_from_db(self):
+        """Load the store of tasks-to-hold from the run DB."""
+        # Note this doesn't need to actually hold the tasks - they're
+        # automatically held at creation via their is_held attribute.
+        for name, cycle, flow_num in (
+            self.db_mgr.pri_dao.select_tasks_to_hold()
+        ):
+            self.hold[(name, get_point(cycle))] = flow_num
+
+    def hold_active_task(
+        self,
+        itask: TaskProxy,
+        flow_num: Optional[int] = None,
+    ) -> bool:
+        """Hold itask if the specified flow_num matches or is None."""
+        if flow_num is not None and flow_num not in itask.flow_nums:
+            # specified flow does not match this task
+            return False
+        if not itask.state_reset(is_held=True):
+            # already held
+            return False
+        self.hold[(itask.tdef.name, itask.point)] = flow_num
+        self._update_stores(itask)
+        return True
+
+    def flag_future_task(
+        self,
+        name: str,
+        point: 'PointBase',
+        flow_num: Optional[int] = None
+    ) -> None:
+        """Flag that we should hold a future task."""
+        self.hold[(name, point)] = flow_num
+        self._update_stores((name, point, True))
```

Comment:
I'm concerned that this means that you can only flag a future task for one flow at a time: 

```console
> cylc hold flows-hold//10100101T0000Z/a --flow 1
Done
> sqlite3 ~/cylc-run/flows-hold/runN/log/db 'SELECT * FROM tasks_to_hold'
a|10100101T0000Z|1
> cylc hold flows-hold//10100101T0000Z/a --flow 2
Done
> sqlite3 ~/cylc-run/flows-hold/runN/log/db 'SELECT * FROM tasks_to_hold'
a|10100101T0000Z|2
```

I would expect to get either
```
a|10100101T0000Z|1,2
# or
a|10100101T0000Z|1
a|10100101T0000Z|2
```

I also think that this means that:
```
> cylc hold flows-hold//10100101T0000Z/a --flow 1 --flow 2
Done
> sqlite3 ~/cylc-run/flows-hold/runN/log/db 'SELECT * FROM tasks_to_hold'
a|10100101T0000Z|2
```
Is not what I as a user would expect: I would expect this to either return an error ("don't use --flow twice please") or set that task as held in both flows as above.

Reply:
I think it would be `--flow=1,2` rather than `--flow=1 --flow=2` but we could extend the CLI to allow this.

Reply:
Ok, in that case it's only the `Reuse the option` that I have a problem with.